presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiPartitionSplitGenerator.java

```diff
@@ -0,0 +1,163 @@
+
+    private final ExtendedHiveMetastore metastore;
+    private final MetastoreContext metastoreContext;
+    private final HudiTableLayoutHandle layout;
+    private final HudiTableHandle table;
+    private final Path tablePath;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final Queue<String> concurrentPartitionQueue;
+    private final String latestInstant;
+    private final HudiSplitWeightProvider splitWeightProvider;
+    private boolean isRunning;
+
+    public HudiPartitionSplitGenerator(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            Queue<String> concurrentPartitionQueue,
+            String latestInstant)
+    {
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.metastoreContext = toMetastoreContext(requireNonNull(session, "session is null"));
+        this.layout = requireNonNull(layout, "layout is null");
+        this.table = layout.getTable();
+        this.tablePath = new Path(table.getPath());
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.concurrentPartitionQueue = requireNonNull(concurrentPartitionQueue, "concurrentPartitionQueue is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+        this.splitWeightProvider = createSplitWeightProvider(requireNonNull(session, "session is null"));
+        this.isRunning = true;
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+
+        while (isRunning || !concurrentPartitionQueue.isEmpty()) {
+            String partitionName = concurrentPartitionQueue.poll();
+
+            if (partitionName != null) {
+                generateSplitsFromPartition(partitionName);
+            }
+        }
+        log.debug("HudiPartitionSplitGenerator %s finishes in %d ms", this, timer.endTimer());
+    }
+
+    public void stopRunning()
+    {
+        this.isRunning = false;
+    }
+
+    private void generateSplitsFromPartition(String partitionName)
+    {
+        HudiPartition hudiPartition = getHudiPartition(metastore, metastoreContext, layout, partitionName);
+        Path partitionPath = new Path(hudiPartition.getStorage().getLocation());
+        String relativePartitionPath = FSUtils.getRelativePartitionPath(tablePath, partitionPath);
+        fsView.getLatestFileSlicesBeforeOrOn(relativePartitionPath, latestInstant, false)
+                .map(fileSlice -> createHudiSplit(table, fileSlice, latestInstant, hudiPartition, splitWeightProvider))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .forEach(asyncQueue::offer);
```

Comment:
`AsyncQueue#offer` returns a ListenableFuture to tell the caller whether the queue is exceeding the outstanding size. We should not discard the return value.

Reply:
thoughts here?

Reply:
Not sure i understood the point here. We are getting the future value in `HudiBackgroundSplitLoader`. Even if the the queue size exceeds, a `notFullSignal` will be returned and the future will be completed eventually when the queue transitions from being full to not full.
https://github.com/prestodb/presto/blob/7da651c695ef2603898185d2a38ced3df33cae2d/presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java#L43-L45